### PR TITLE
flask healthz endpoints added

### LINKED ISSUE #269: Integrate Flask Healthz into datagrepper
Use flask-healthz to provide endpoints for readiness probes:

https://github.com/fedora-infra/flask-healthz

### PR DESCRIPTION
**Pull request by james02135**
Signed-off-by: james02135 <james02135@hotmail.com>

### CODE REVIEW

File: datagrepper/app.py [hunk 1]
```diff
@@ -773,3 +778,14 @@ def internal_error(error):
...
     )
+
+
+def liveness():
+    pass
+
+
+def readiness():
+    try:
+        print("everything looks good")
```
Comment (abompard) [inline datagrepper/app.py]: It would be more useful to try a database query, for example something like `SELECT 1`. It would ensure that the database connection is actually working.

File: datagrepper/app.py [hunk 2]
```diff
@@ -773,3 +778,14 @@ def internal_error(error):
...
+    except Exception:
+        raise HealthError("Can't connect to the file")
```
Comment (abompard) [inline datagrepper/app.py]: Why this error message?

**Review by abompard — CHANGES_REQUESTED** [review]

File: datagrepper/app.py [hunk 3]
```diff
@@ -776,3 +781,14 @@ def internal_error(error):
...
     )
+
+
+def liveness():
+    pass
+
+
+def readiness():
+    try:
+        datagrepper.ping()
```
Comment (abompard) [inline datagrepper/app.py]: Where does this "ping" function come from? I can't find it.

File: datagrepper/app.py [hunk 4]
```diff
@@ -776,3 +781,14 @@ def internal_error(error):
...
+    except Exception:
+        raise HealthError("Can't connect to the Datagrepper server")
```
Comment (abompard) [inline datagrepper/app.py]: I don't know what the `ping()` function does but I suspect it connects to the database server, not the "Datagrepper" server.

File: datagrepper/app.py (hunk 3 above)
Comment (james02135) [reply datagrepper/app.py]: Just so I understand this better, in pseudo-code it would look something like this - 

def readiness():
       try:
             connect_to_the_database()
             query = "SELECT 1 FROM datagrepper;"
       except Exception:
              raise HealthError("Can't connect to the database")      
      return (query)

Comment (abompard) [reply datagrepper/app.py]: Yeah, something like that. The DB connection should already have been established, it is done at the start of the application (see `dm.init()` in `app.py`). (it is, by the way, not the best way to do it, but that's a different issue)

What you have to do in this function is make sure that you can actually query the database by executing the SQL query.
Also, the query can just be `SELECT 1`, no need for the `FROM tablename`, we're just trying to check that the DB server is responding here, not running actual data queries.
And you don't need to return the query.

Comment (james02135) [reply datagrepper/app.py]: Have to fix some of the tests, but the code should be closer to what we want

File: pyproject.toml [hunk 5]
```diff
@@ -85,6 +85,8 @@ psycopg2 = "^2.9.1"
 cffi = "^1.14.6"
 gunicorn = {version = "^20.0", optional = true}
 SQLAlchemy = "<1.4.23"
+flask-healthz = "^0.0.1"
+cryptography = "3.4.7"
```
Comment (abompard) [inline pyproject.toml]: I think you need to run `poetry update` so that the pinning of `cryptography` is recorded in the `poetry.lock` file. The addition of `flask-healthz` was automatically recorded when you used the command `poetry add`, but manual changes to `pyproject.toml` must be followed by a `poetry lock`.

File: tests/test_api.py [hunk 6]
```diff
@@ -146,6 +147,24 @@ def test_chart_line(self, grep):
         resp = self.client.get("/charts/line")
         self.assertEqual(resp.status_code, 200)
         self.assertEqual(resp.mimetype, "image/svg+xml")
-        self.assertIn(
-            b'<svg xmlns:xlink="http://www.w3.org/1999/xlink', resp.get_data()
-        )
```
Comment (abompard) [inline tests/test_api.py:151]: Why remove this?

**Review by abompard — CHANGES_REQUESTED** [review]

File: tests/test_api.py [hunk 7]
```diff
@@ -149,3 +150,24 @@ def test_chart_line(self, grep):
...
+
+    def test_healthz_liveness(self):
+        """Test the /healthz/live check endpoint"""
+        result = self.client.get("/healthz/live")
+        assert result.status_code == 200
+        assert result.data == b"OK\n"
+
+    def test_healthz_readiness_ok(self):
+        """Test the /healthz/ready check endpoint"""
+        with mock.patch("datagrepper.app.readiness"):
```
Comment (abompard) [inline tests/test_api.py]: We should not mock the code that we're testing, otherwise we end up testing the mock and not the real function. In this case, `datagrepper.app.readiness` is the function that this testcase is testing, no?

File: tests/test_api.py [hunk 8]
```diff
@@ -149,3 +150,24 @@ def test_chart_line(self, grep):
...
+            result = self.client.get("/healthz/ready")
+        assert result.status_code == 200
+        assert result.data == b"OK\n"
+
+    def test_healthz_readiness_not_ok(self):
+        """Test the /healthz/ready check endpoint when not ready"""
+        with mock.patch("datagrepper.app.index") as greppindex:
```
Comment (abompard) [inline tests/test_api.py]: Why mock `datagrepper.app.index`? It does not seem related to the readiness function.

**Review by abompard — CHANGES_REQUESTED** [review]

File: tests/test_api.py [hunk 9]
```diff
@@ -4,6 +4,7 @@
 import json
 import unittest
 from datetime import datetime, timedelta, timezone
+from unittest import mock
```
Comment (StephenCoady) [inline tests/test_api.py]: If you use the decorator like we spoke about yesterday you won't need this import.

File: tests/test_api.py [hunk 10]
```diff
@@ -149,3 +150,23 @@ def test_chart_line(self, grep):
...
+
+    def test_healthz_readiness_ok(self):
+        """Test the /healthz/ready check endpoint"""
+        result = self.client.get("/healthz/ready")
+        assert result.status_code == 200
+        assert result.data == b"OK\n"
+
+    def test_healthz_readiness_not_ok(self):
+        """Test the /healthz/ready check endpoint when not ready"""
+        with mock.patch("datagrepper.app.dm.session") as grepsesh:
```
Comment (StephenCoady) [inline tests/test_api.py]: You're mocking the wrong function here again. You don't use `dm.session` in `readiness()`.

Comment (StephenCoady) [inline tests/test_api.py]: you should use the decorator as well to keep in line with the other test's code style.

**Review by StephenCoady — CHANGES_REQUESTED** [review]

**Review by abompard — APPROVED** [review]
